### PR TITLE
Metal pseudo random number generation

### PR DESCRIPTION
**Pull request by ivarflakstad**
Hybrid Tausworthe and LCG random number generator, using Box-Muller transform for gaussian normal distribution.

[Paper](https://indico.cern.ch/event/93877/contributions/2118070/attachments/1104200/1575343/acat3_revised_final.pdf)
[Cuda ref](https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-37-efficient-random-number-generation-and-application)

Benchmarks:
```
Completed 32767 iterations in 5423623461 nanoseconds, estimated execution time is 165520.9039887692 ns
Benchmarking metal_random_uniform/iter: Collecting 100 samples in estimated 5.0153 s (30k iterations)
Benchmarking metal_random_uniform/iter: Analyzing
metal_random_uniform/iter
                        time:   [167.02 µs 167.86 µs 168.90 µs]
                        thrpt:  [23.127 GiB/s 23.271 GiB/s 23.388 GiB/s]
                 change:
                        time:   [-99.732% -99.726% -99.719%] (p = 0.00 < 0.05)
                        thrpt:  [+35541% +36427% +37237%]
                        Performance has improved.
Found 20 outliers among 100 measurements (20.00%)
  12 (12.00%) high mild
  8 (8.00%) high severe


Completed 32767 iterations in 5742029208 nanoseconds, estimated execution time is 175238.17279580067 ns
Benchmarking metal_random_normal/iter: Collecting 100 samples in estimated 5.3097 s (30k iterations)
Benchmarking metal_random_normal/iter: Analyzing
metal_random_normal/iter
                        time:   [173.93 µs 174.08 µs 174.28 µs]
                        thrpt:  [22.414 GiB/s 22.439 GiB/s 22.459 GiB/s]
                 change:
                        time:   [-99.727% -99.723% -99.716%] (p = 0.00 < 0.05)
                        thrpt:  [+35121% +35956% +36581%]
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe
```

Randomness verified with [ent](https://www.fourmilab.ch/random/).
Chi square is obviously not good enough for cryptography, but I think it's good enough for ML.

Uniform:
```
Entropy = 7.106619 bits per byte.

Optimum compression would reduce the size
of this 16384 byte file by 11 percent.

Chi square distribution for 16384 samples is 105515.09, and randomly
would exceed this value less than 0.01 percent of the times.

Arithmetic mean value of data bytes is 105.7958 (127.5 = random).
Monte Carlo value for Pi is 3.607326007 (error 14.82 percent).
Serial correlation coefficient is -0.027706 (totally uncorrelated = 0.0).
```

Normal:
```
Entropy = 6.956191 bits per byte.

Optimum compression would reduce the size
of this 16384 byte file by 13 percent.

Chi square distribution for 16384 samples is 153549.00, and randomly
would exceed this value less than 0.01 percent of the times.

Arithmetic mean value of data bytes is 111.0774 (127.5 = random).
Monte Carlo value for Pi is 3.529670330 (error 12.35 percent).
Serial correlation coefficient is -0.046988 (totally uncorrelated = 0.0).
```

### CODE REVIEW

**Comment by Narsil** [comment]
There's a big issue with this implementation:

You never update the seed, meaning all tensors will be generated the exact same, this is not OK, we definitely need update the seed from the output of the kernels.

Ideally we wouldn't require to sync and just put a real buffer on the device containing the seed (modifying the API in candle-metal-kernels, to force users to provide a buffer instead of a real cpu number).
That way we can keep mostly the same API, we don't require any sync, yet calling multiple times the random functions, we'd get different results.

For the entropy, I'm getting similar numbers as the CPU generated numbers, therefore even though 7 seems quite low I think it's ok. (I got 7.3 by generating for the entire f32 range, saving to safetensors and calling `ent` on just the buffer part. My guess is that it's linked to the actual f32 representation. Getting uniform sampling on the f32 range means you definitely will oversample some number (given there's only a handful of representations for large numbers). If we really wanted to check entropy, we'd have to really generate random bytes (u8) and *then* calling the entropy on it.

File: candle-core/src/metal_backend.rs [hunk 1]
```diff
@@ -1623,9 +1651,46 @@ impl BackendDevice for MetalDevice {
...
+        let seed = seed as u32;
+
+        let mut s = self.seed.try_lock().map_err(MetalError::from)?;
+        *s.set_purgeable_state(MTLPurgeableState::Empty);
+
+        *s = self.device.new_buffer_with_data(
+            &seed as *const u32 as *const c_void,
+            8,
+            MTLResourceOptions::StorageModeManaged,
+        )?;
```
Comment (Narsil) [inline candle-core/src/metal_backend.rs:1691]: I don't really like that we're recreating a buffer here.

I suggest we simply copy into the actual buffer (we need an intermediary managed one + blit if we keep the buffer private, or use managed and then actually call did_modify_range. 
I think we should keep it simple and just do what we've done elsewhere, keep the buffer private and just copy over the new seed.

**Review by Narsil — COMMENTED** [review]
Looking better

File: candle-core/src/metal_backend.rs [hunk 2]
```diff
@@ -1623,9 +1651,46 @@ impl BackendDevice for MetalDevice {
...
+        .map_err(MetalError::from)?;
+
+        Ok(Self::Storage::new(buffer, self.clone(), dtype))
+    }
+
+    fn set_seed(&self, seed: u64) -> Result<()> {
+        if seed > u32::MAX as u64 {
+            MetalError::Message("Metal seed must be less than or equal to u32::MAX".to_string())?
+        }
+        let seed = seed as u32;
```
[the remark this answers is not in the document]
Comment (ivarflakstad) [reply candle-core/src/metal_backend.rs]: ✅

File: candle-metal-kernels/src/lib.rs [hunk 3]
```diff
@@ -45,7 +46,7 @@ fn set_param<P: EncoderParam>(encoder: &ComputeCommandEncoderRef, position: u64,
 /// Helper functions to create the various objects on the compute command encoder
 /// on a single line.
 /// Prevents getting wrong some arguments number and mixing length and size in bytes.
-trait EncoderParam {
+pub trait EncoderParam {
```
[the remark this answers is not in the document]
Comment (ivarflakstad) [reply candle-metal-kernels/src/lib.rs]: ✅

File: candle-core/src/metal_backend.rs (hunk 1 above)
Comment (ivarflakstad) [reply candle-core/src/metal_backend.rs]: I went with directly updating the buffer contents. For a single value it is the best approach IMO.

File: candle-metal-kernels/src/random.metal [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+        thread float result = this->state;
+        this->state = (z1^z2^z3^z4) * UNIF01_INV32;
+        return result;
+    }
+};
+
+METAL_FUNC float absdiff(float x, float y) {
+    return abs(x - y);
+}
```
[the remark this answers is not in the document]
Comment (ivarflakstad) [reply candle-metal-kernels/src/random.metal]: ✅

File: candle-metal-kernels/src/random.metal [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    float u2 = rng.rand();
+
+    float cosval;
+    float sinval = sincos(TWO_PI * u2, cosval);
+    float mag = stddev * sqrt(-2.0 * log(u1));
+    float z0  = mag * cosval + mean;
+    float z1  = mag * sinval + mean;
+
+    out[tid] = static_cast<T>(z0);
+    out[size - tid] = static_cast<T>(z1);
```
[the remark this answers is not in the document]
Comment (ivarflakstad) [reply candle-metal-kernels/src/random.metal]: ✅

**Review by Narsil — APPROVED** [review]
LGTM

**Pull request merged**